core/src/main/java/com/datastrato/gravitino/storage/relational/mapper/SecurableObjectMapper.java

```diff
@@ -0,0 +1,82 @@
+ */
+package com.datastrato.gravitino.storage.relational.mapper;
+
+import com.datastrato.gravitino.storage.relational.po.SecurableObjectPO;
+import java.util.List;
+import org.apache.ibatis.annotations.Delete;
+import org.apache.ibatis.annotations.Insert;
+import org.apache.ibatis.annotations.Param;
+import org.apache.ibatis.annotations.Select;
+import org.apache.ibatis.annotations.Update;
+
+/**
+ * A MyBatis Mapper for table meta operation SQLs.
+ *
+ * <p>This interface class is a specification defined by MyBatis. It requires this interface class
+ * to identify the corresponding SQLs for execution. We can write SQLs in an additional XML file, or
+ * write SQLs with annotations in this interface Mapper. See: <a
+ * href="https://mybatis.org/mybatis-3/getting-started.html"></a>
+ */
+public interface SecurableObjectMapper {
+
+  String SECURABLE_OBJECT_TABLE_NAME = "role_meta_securable_object";
+  String ROLE_TABLE_NAME = "role_meta";
+
+  @Insert({
+    "<script>",
+    "INSERT INTO "
+        + SECURABLE_OBJECT_TABLE_NAME
+        + "(role_id, entity_id, type, privilege_names, privilege_conditions, "
+        + " current_version, last_version, deleted_at)"
+        + " VALUES ",
+    "<foreach collection='securableObjects' item='item' separator=','>",
+    "(#{item.roleId},"
+        + " #{item.entityId},"
+        + " #{item.type},"
+        + " #{item.privilegeNames},"
+        + " #{item.privilegeConditions},"
+        + " #{item.currentVersion},"
+        + " #{item.lastVersion},"
+        + " #{item.deletedAt})",
+    "</foreach>",
+    "</script>"
+  })
+  void batchInsertSecurableObjects(
+      @Param("securableObjects") List<SecurableObjectPO> securableObjectPOS);
+
+  @Update(
+      "UPDATE "
+          + SECURABLE_OBJECT_TABLE_NAME
+          + " SET deleted_at = UNIX_TIMESTAMP(CURRENT_TIMESTAMP(3)) * 1000.0"
+          + " WHERE role_id = #{roleId} AND deleted_at = 0")
+  void softDeleteSecurableObjectsByRoleId(@Param("roleId") Long roleId);
+
+  @Update(
+      "UPDATE "
+          + SECURABLE_OBJECT_TABLE_NAME
+          + " ob SET ob.deleted_at = UNIX_TIMESTAMP(CURRENT_TIMESTAMP(3)) * 1000.0"
+          + " where exists ( select * from "
+          + ROLE_TABLE_NAME
+          + " ro WHERE ro.metalake_id = #{metalakeId} AND ro.role_id = ob.role_id AND ro.deleted_at = 0) AND ob.deleted_at = 0")
+  void softDeleteRoleMetasByMetalakeId(@Param("metalakeId") Long metalakeId);
+
+  @Select(
+      "SELECT role_id as roleId, entity_id as entityId,"
+          + " type as type, privilege_names as privilegeNames,"
+          + " privilege_conditions as privilegeConditions, current_version as currentVersion,"
+          + " last_version as lastVersion, deleted_at as deletedAt"
+          + " FROM "
+          + SECURABLE_OBJECT_TABLE_NAME
+          + " WHERE role_id = #{roleId} AND deleted_at = 0")
+  List<SecurableObjectPO> listSecurableObjectsByRoleId(@Param("roleId") Long roleId);
```

Comment:
Does this mean that if this securable object binds to a entity like table, which has metalake_id, catalog_id, schema_id, and table_id, here we will insert 4 rows for this securable object with the same privilege, right?

So if user has another securable like catalog has the same privilege, it will insert metalake_id, catalog_id 2 rows. These two rows are exactly the same as above 2 rows in 4, how do you distinguish which row maps to which securable object?

Reply:
This can be a problem when you delete a securable object, for example, if you want to delete the case 1, instead of deleting 4 rows, you will probably delete 6 rows.

Reply:
No, we just insert one row for every entity id.